breacharbiter.go

```diff
@@ -436,62 +449,91 @@ func (b *breachArbiter) waitForSpendEvent(breachInfo *retributionInfo,
 		// channel before ranging over its content.
 		close(allSpends)
 
-		doneOutputs := make(map[int]struct{})
+		// Gather all detected spends and return them.
+		var spends []spend
 		for s := range allSpends {
 			breachedOutput := &inputs[s.index]
 			delete(spendNtfns, breachedOutput.outpoint)
 
-			switch breachedOutput.witnessType {
-			case input.HtlcAcceptedRevoke:
-				fallthrough
-			case input.HtlcOfferedRevoke:
-				brarLog.Infof("Spend on second-level"+
-					"%s(%v) for ChannelPoint(%v) "+
-					"transitions to second-level output",
-					breachedOutput.witnessType,
-					breachedOutput.outpoint,
-					breachInfo.chanPoint)
+			spends = append(spends, s)
+		}
 
-				// In this case we'll morph our initial revoke
-				// spend to instead point to the second level
-				// output, and update the sign descriptor in the
-				// process.
-				convertToSecondLevelRevoke(
-					breachedOutput, breachInfo, s.detail,
-				)
+		return spends, nil
 
-				continue
+	case <-b.quit:
+		return nil, errBrarShuttingDown
+	}
+}
+
+// updateBreachInfo mutates the passed breachInfo by removing or converting any
+// outputs among the spends.
+func updateBreachInfo(breachInfo *retributionInfo, spends []spend) {
+	inputs := breachInfo.breachedOutputs
+	doneOutputs := make(map[int]struct{})
+
+	for _, s := range spends {
+		breachedOutput := &inputs[s.index]
+		txIn := s.detail.SpendingTx.TxIn[s.detail.SpenderInputIndex]
+
+		switch breachedOutput.witnessType {
+		case input.HtlcAcceptedRevoke:
+			fallthrough
+		case input.HtlcOfferedRevoke:
+			signDesc := breachedOutput.signDesc
+			revokeKey := input.DeriveRevocationPubkey(
+				signDesc.KeyDesc.PubKey,
+				signDesc.DoubleTweak.PubKey(),
+			).SerializeCompressed()
+
+			// If the HTLC output was spent using the revocation
+			// key, it is our own spend, and we can forget the
+			// output. Otherwise it has been taken to the second
+			// level.
+			if len(txIn.Witness) == 3 &&
```

Comment:
cc @Crypt-iQ

Reply:
Looks good 👍

Reply:
Extracted into method: https://github.com/lightningnetwork/lnd/pull/5036/commits/161c1b9eeac778657417f9c46d9bc01ad9df48ac

Could add a simple unit test, that's true.